content/blog/2019/02/2019-02-12-jenkins-alexa-voice-controlled-cicd.adoc

```diff
@@ -0,0 +1,115 @@
+* Alexa Skills Kit (ASK) - ASK is the "SDK" (Software Development Kit) that allows developers to build custom skills for Alexa. 
+* Alexa Developer Portal - An Alexa skill includes a voice user interface, or VUI, to understand user intents, and a back-end cloud service to process intents by telling Alexa how to respond. The VUI and the integration with the back-end service is setup and configured through the Alexa Developer Portal.
+* AWS Lambda - A chunk of code that runs in the cloud. Developers can run their code without having to provision or manage servers. Applications created with AWS Lambda are considered to be serverless. Lambda supports several popular languages like Python, Java, Node.js, Go, C#, etc.
+* GitHub - A version control system for the Java project source code. 
+* Jenkins on EC2 - I use Jenkins to build, test, and deploy my Java Application Programming Interface (API). Elastic Cloud Computer (EC2) is the virtual server where Jenkins is installed. Jenkins works alongside several other tools:
+. Maven - A build automation tool for Java projects.
+. Junit - A testing framework for Java projects.
+. AWS Command Line Interface (CLI) - This is a command line tool that allows developers to access their Amazon Web Services (AWS) account.
+. Blue Ocean - This is a plugin for Jenkins that provides an easy to use interface to create and monitor Jenkins pipelines. 
+* AWS Elastic Beanstalk - This is an orchestration service that allows developers to deploy and manage web applications in the AWS cloud. 
+* Postman - This is an HTTP client for testing APIs and web services.
+
+== Voice Interaction Model
+
+The Voice User Interface (VUI) describes the overall conversational flow and is setup via the Alexa Developer Console. 
+
+image::/images/post-images/2019-02-12-alexa/alexa_dev_console.png[Invocation Name Setup Via Alexa Developer Console, role=center]
+
+A few important components of the VUI are the Invocation Name (how users launch your skill) and the Intents (phrases a user says to "talk to" or interact with your skill). 
+
+image::/images/post-images/2019-02-12-alexa/alexa_dev_console_deploy.png[Utterances for DeployCodeIntent Via Alexa Developer Console, role=center]
+
+Specifically, the "DeployCodeIntent" is invoked when a user says one of several phrases (e.g. run jenkins pipeline, run jenkins job, deploy the code, deploy code, or deploy ) or a variation of the phrase like, "deploy my code".
+
+image::/images/post-images/2019-02-12-alexa/alexa_dev_console_endpoint.png[Backend Fulfillment Logic - Endpoint Via Alexa Developer Console, role=center]
+
+The endpoint is the destination where the skill requests are sent for fulfillment. 
+In this case, the backend logic is an AWS Lambda authored in Python. 
+The business logic in the Lambda uses the Jenkins API to invoke the job. 
```

Comment:
Since it's authored in Python, it's ambiguous whether this uses [jenkinsapi](https://pypi.org/project/jenkinsapi/), an independent Python library, or the regular Jenkins API. Should be clarified, perhaps with a link, if it's the former.

Reply:
I'm using the Jenkins remote access API to trigger the job remotely. The URL used takes the following format: **_jenkins_url/job/job_name/build_**. I'm using BASIC authentication and passing in a Jenkins Crumb in the HTTP request header for CSRF protection.

Reply:
Since 2.96, you don't need a crumb if you use the API token instead of a password. That should make it easier.

Ideally these details should make their way into the blog post.

Reply:
The solution I've implemented doesn't use an API token. However, I'm fine with a "side note" being added that mentions alternative solutions.

Reply:
Updated pr with the additional information.